File: creme/stats/covariance.py
from . import base


class Covariance(base.Bivariate):
    """Running covariance.

    Parameters:
        ddof: Delta Degrees of Freedom.

    Example:

        >>> from creme import stats

        >>> x = [-2.1,  -1,  4.3]
        >>> y = [   3, 1.1, 0.12]

        >>> cov = stats.Covariance()

        >>> for xi, yi in zip(x, y):
        ...     print(cov.update(xi, yi).get())
        0.0
        -1.044999
        -4.286

    References:
        1. [Wikipedia article on algorithms for calculating variance](https://www.wikiwand.com/en/Algorithms_for_calculating_variance#/Covariance)

    """

    def __init__(self, ddof=1):
        self.ddof = ddof
        self.mean_x = 0
        self.mean_y = 0
        self.n = 0
        self.cov = 0

    def update(self, x, y):
        self.n += 1
        dx = x - self.mean_x
        self.mean_x += dx / self.n
        self.mean_y += (y - self.mean_y) / self.n
        self.cov += (dx * (y - self.mean_y) - self.cov) / max(1, self.n - self.ddof)
        return self

    def get(self):
        return self.cov

File: _watermill/src/covariance.rs
use num::{Float, FromPrimitive};
use std::ops::{AddAssign, SubAssign};

use crate::mean::Mean;
use crate::stats::{Bivariate, Univariate};
use serde::{Deserialize, Serialize};
/// Running Covariance.
/// # Examples
/// ```
/// use watermill::covariance::Covariance;
/// use watermill::stats::Bivariate;
/// let mut running_cov: Covariance<f64> = Covariance::default();
/// let x: Vec<f64> = vec![-2.1,  -1.,  4.3];
/// let y: Vec<f64> = vec![3., 1.1, 0.12];
/// for (xi, yi) in x.iter().zip(y.iter()){
///     running_cov.update(*xi,*yi);
/// }
/// assert_eq!(running_cov.get(), -4.286);
/// ```
/// # References
/// [^1]: [Wikipedia article on algorithms for calculating variance](https://www.wikiwand.com/en/Algorithms_for_calculating_variance#/Covariance)
///
/// [^2]: Schubert, E. and Gertz, M., 2018, July. Numerically stable parallel computation of (co-) variance. In Proceedings of the 30th International Conference on Scientific and Statistical Database Management (pp. 1-12).
#[derive(Clone, Copy, Debug, Serialize, Deserialize)]
pub struct Covariance<F: Float + FromPrimitive + AddAssign + SubAssign> {
    pub ddof: u32,
    pub mean_x: Mean<F>,
    pub mean_y: Mean<F>,
    c: F,
    pub cov: F,
}
impl<F: Float + FromPrimitive + AddAssign + SubAssign> Covariance<F> {
    pub fn new(ddof: u32) -> Self {
        Self {
            mean_x: Mean::new(),
            mean_y: Mean::new(),
            ddof,
            c: F::from_f64(0.).unwrap(),
            cov: F::from_f64(0.).unwrap(),
        }
    }
}

impl<F> Default for Covariance<F>
where
    F: Float + FromPrimitive + AddAssign + SubAssign,
{
    fn default() -> Self {
        Self {
            ddof: 1,
            mean_x: Mean::new(),
            mean_y: Mean::new(),
            c: F::from_f64(0.).unwrap(),
            cov: F::from_f64(0.).unwrap(),
        }
    }
}

impl<F: Float + FromPrimitive + AddAssign + SubAssign> Bivariate<F> for Covariance<F> {
    fn update(&mut self, x: F, y: F) {
        let dx = x - self.mean_x.get();
        self.mean_x.update(x);
        self.mean_y.update(y);
        self.c += dx * (y - self.mean_y.get());
        self.cov = self.c
            / (F::from_f64(1.)
                .unwrap()
                .max(self.mean_x.n.get() - F::from_u32(self.ddof).unwrap()));
    }
    fn get(&self) -> F {
        self.cov
    }
}
